### PR TITLE
Adding Speaker Status and Speaker Title in Events Form

### LINKED ISSUE #7820: [BUG] Events no more show the speaker status and topic
### Description

The user status and topic are not being displayed on the event tile on the events page (https://linkfree.io/events). They used to appear when the user added the following properties:

```json
  "userStatus": "speaking",
  "speakingTopic": "Open Source"
```

### Screenshots

Screenshot 2023-06-28 at 1 42 34 PM

### Additional information

We need to be careful while fixing this bug because now we group events, and what if two users are speaking at the same conference?

### PR DESCRIPTION
**Pull request by akash19coder**
## Fixes Issue
Closes #7820 

## Changes proposed
 **1. Added two inputs field in the events form namely: Are you speaking? and What topic are you speaking on?**

Screenshot (23)

** 2. User profile to display speaking information** 
Screenshot (22)

## Check List (Check all the applicable boxes) 

- [x] My code follows the code style of this project.
- [ ] My change requires changes to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] All new and existing tests passed.
- [x] This PR does not contain plagiarized content.
- [x] The title of my pull request is a short description of the requested changes.

## Note to reviewers

<a href="https://gitpod.io/#https://github.com/EddieHubCommunity/LinkFree/pull/8006"></a>

### CODE REVIEW

**Review by eddiejaoude — CHANGES_REQUESTED** [review]
Thank you for the contribution 👍 

A few things I noticed:
- too many strings being used, we should move these to config
- json events have this info and needs to be loaded into existing events

**Comment by akash19coder** [comment]
> * json events have this info and needs to be loaded into existing events

eddie what json events are we referring to could you please give me some more context. I actually didn't understand

**Comment by eddiejaoude** [comment]
> > * json events have this info and needs to be loaded into existing events
> 
> eddie what json events are we referring to could you please give me some more context. I actually didn't understand

The reload script will need updating for existing events managed by the json files https://github.com/EddieHubCommunity/LinkFree/blob/main/pages/api/system/reload.js

File: models/Profile/Event.js [hunk 1]
```diff
@@ -34,6 +34,11 @@ const EventSchema = new Schema({
   price: {
     startingFrom: Number,
   },
+  userStatus: {
+    type: String,
+    enum: ["speaking", "not-speaking"],
```
Comment (eddiejaoude) [inline models/Profile/Event.js:39]: I think the options should be `speaking`, `attending`

File: pages/account/manage/event/[[...data]].js [hunk 2]
```diff
@@ -69,6 +70,14 @@ export default function ManageEvent({ BASE_URL, event }) {
   const [endDate, setEndDate] = useState(
     event.date?.end && formatDate(event.date?.end)
   );
+  const DropDownOptions = ["Non-speaking","Speaking"];
```
Comment (eddiejaoude) [inline pages/account/manage/event/[[...data]].js:73]: Duplicate again, please set in config and reuse

File: pages/account/manage/event/[[...data]].js [hunk 3]
```diff
@@ -69,6 +70,14 @@ export default function ManageEvent({ BASE_URL, event }) {
...
+  const [userStatus, setuserStatus] = useState(event.userStatus || "");
+  const [speakingTopic, setspeakingTopic] = useState(event.speakingTopic || "Your speaking topic");
```
Comment (eddiejaoude) [inline pages/account/manage/event/[[...data]].js:75]: Default should be an empty string, use html placeholder for suggestions

File: pages/account/manage/event/[[...data]].js [hunk 4]
```diff
@@ -238,6 +249,35 @@ export default function ManageEvent({ BASE_URL, event }) {
                       For example: In <i>DD / MM / YYYY, HH:MM</i>
                     </p>
                   </div>
+                  <div className="mt-1 sm:col-span-2 sm:mt-0">
+                  <div className="mt-1">
+                        <Select
```
Comment (eddiejaoude) [inline pages/account/manage/event/[[...data]].js:254]: Formatting doesn't look correct here, Prettier VScode plugin will fix

File: models/Profile/Event.js [hunk 5]
```diff
@@ -34,6 +34,11 @@ const EventSchema = new Schema({
...
+  },
+  speakingTopic: String,
```
Comment (eddiejaoude) [inline models/Profile/Event.js:41]: This needs a max, maybe 256?

File: pages/account/manage/event/[[...data]].js [hunk 6]
```diff
@@ -238,6 +249,35 @@ export default function ManageEvent({ BASE_URL, event }) {
...
+                      </div>
+                      <p className="text-sm text-primary-low-medium">
+                        Select Speaking or Non-speaking
+                      </p>
+                    </div>
+                    {userStatus === "Speaking" && (
+                      <div className="mt-1 sm:col-span-2 sm:mt-0">
+                        <Input
+                          type="text"
+                          min="0"
```
Comment (eddiejaoude) [inline pages/account/manage/event/[[...data]].js:270]: min 0 is not recommended, if required, used `required`

also a `max` is required, this should match the model max

File: pages/account/manage/event/[[...data]].js [hunk 7]
```diff
@@ -144,7 +155,7 @@ export default function ManageEvent({ BASE_URL, event }) {
   return (
     <>
       <PageHead
-        title="Manage Milstone"
+        title="Manage Milestone"
```
Comment (eddiejaoude) [inline pages/account/manage/event/[[...data]].js:158]: Good spot 👍

File: config/app.json [hunk 8]
```diff
@@ -6,6 +6,7 @@
     "inline",
     "column-2"
   ],
+  "userStatus":"Speaking",
```
Comment (eddiejaoude) [inline config/app.json:9]: This is not clear, it just seems randomly added, I think it needs to go into an `event` object

Comment (akash19coder) [reply config/app.json:9]: This is how I have represented, is it fine? 
```
"layouts": [
    "classic",
    "inline",
    "column-2"
  ],
  "events":{
    "userStatus":["Attending","Speaking"]
  },
  "isr": {
    "homepage": {
      "revalidateSeconds": 900,
      "description": "15 * 60 = 15 minutes"
    },
```

File: pages/account/manage/event/[[...data]].js (hunk 6 above)
Comment (akash19coder) [reply pages/account/manage/event/[[...data]].js:270]: Here is the change:
```
                        <Input
                          type="text"
                          name="topic"
                          label="What topic are you speaking on ?"
                          placeholder= "Your speaking topic"
                          onChange={(e) => setspeakingTopic(e.target.value)}
                          value={speakingTopic}
                          required
                          maxLength="256"
                        />
```

File: models/Profile/Event.js (hunk 5 above)
Comment (akash19coder) [reply models/Profile/Event.js:41]: Here:
```
speakingTopic: {
    type: String,
    max: 256
  }
```

File: pages/account/manage/event/[[...data]].js (hunk 3 above)
Comment (akash19coder) [reply pages/account/manage/event/[[...data]].js:75]: Added default to be an empty string:

`const [speakingTopic, setspeakingTopic] = useState(event.speakingTopic || "");`

Added placeholder in the component:
```
 <Input
   type="text"
   name="topic"
   label="What topic are you speaking on ?"
   placeholder= "Your speaking topic"
   onChange={(e) => setspeakingTopic(e.target.value)}
   value={speakingTopic}
   required
   maxLength="256"
  />
```

File: pages/account/manage/event/[[...data]].js (hunk 2 above)
Comment (akash19coder) [reply pages/account/manage/event/[[...data]].js:73]: Removed duplicacy:
```
  const [endDate, setEndDate] = useState(
    event.date?.end && formatDate(event.date?.end)
  );
  const DropDownOptions = config.events.userStatus;
```

File: components/event/EventCard.js [hunk 9]
```diff
@@ -75,11 +76,11 @@ export default function EventCard({ manage, event, usernames }) {
...
                   )}
                 </div>
-                {event.userStatus && (
+                {event.userStatus == config.userStatus && (
                   <div className="text-primary-medium-low dark:text-primary-low-medium italic hidden lg:block">
                     {event.userStatus}
-                    {event.userStatus == "speaking" && " at "} this event
-                    {event.userStatus == "speaking" && event?.speakingTopic && (
+                    {event.userStatus == config.userStatus && " at "} this event
+                    {event.userStatus == config.userStatus && event?.speakingTopic && (
```
Comment (akash19coder) [inline components/event/EventCard.js:83]: Based on new config updation i.e : 
 ```
 "layouts": [
    "classic",
    "inline",
    "column-2"
  ],
  "events":{
    "userStatus":["Attending","Speaking"]
  },
```
Here is the change, what do you think Eddie?
```
             {event.userStatus == config.events.userStatus[1] && (
                  <div className="text-primary-medium-low dark:text-primary-low-medium italic hidden lg:block">
                    {event.userStatus}
                    {event.userStatus == config.events.userStatus[1] && " at "} this event
                    {event.userStatus == config.events.userStatus[1] && event?.speakingTopic && (
                      <>
                        {" "}
                        on <b>{event.speakingTopic}</b>
                      </>
                    )}
                  </div>
```

**Comment by akash19coder** [comment]
If all the changes are alright will commit the changes and push. 😄

File: config/app.json (hunk 8 above)
Comment (eddiejaoude) [reply config/app.json:9]: Yes this is better 👍 thank you

File: pages/account/manage/event/[[...data]].js (hunk 6 above)
Comment (eddiejaoude) [reply pages/account/manage/event/[[...data]].js:270]: Is it required though? What if they are not speaking?

File: models/Profile/Event.js (hunk 5 above)
Comment (eddiejaoude) [reply models/Profile/Event.js:41]: Please make the change in your PR, it is easier to review

**Comment by eddiejaoude** [comment]
Thank you for the replies, please can you commit the changes so it is easier to review 

Also, can you resolve conflicts, thank you 👍

File: config/app.json [hunk 10]
```diff
@@ -6,6 +6,9 @@
     "inline",
     "column-2"
   ],
+  "events":{
+    "userStatus":["Attending","Speaking"]
```
Comment (eddiejaoude) [inline config/app.json:10]: Looking at the usage code from here, it is still not readable or maintainable

```js
event.userStatus == config.events.userStatus[1]
```

So I suggest using an object not an array

```suggestion
    "userStatus": { attending: "Attending", speaking: "Speaking"}
```

Then the usage will become

```js
event.userStatus == config.events.speaking
```

File: models/Profile/Event.js (hunk 5 above)
Comment (eddiejaoude) [reply models/Profile/Event.js:43]: You marked this as resolved but the changes haven't been made, any reason why?

**Comment by eddiejaoude** [comment]
@akash19coder you keep merging main into your branch, however you haven't addressed some comments still but you keep chasing me?

**Comment by akash19coder** [comment]
I was so desperate for the PR to merge that I overlooked the mistake. I am sorry.

> @akash19coder you keep merging main into your branch, however you haven't addressed some comments still but you keep chasing me?

I was so desperate for the PR to merge that I overlooked the mistake. my bad. 🙂

File: pages/account/manage/event/[[...data]].js (hunk 6 above)
Comment (akash19coder) [reply pages/account/manage/event/[[...data]].js]: If the user selects the speaking option it only appears then else doesn't appear

File: models/Profile/Event.js (hunk 5 above)
Comment (akash19coder) [reply models/Profile/Event.js]: I have added the field in a new commit

**Comment by akash19coder** [comment]
@eddiejaoude I haven't made any changes in UserEvent.js or event.js file. However, when I create events it says the event was created successfully but doesn't load the event in the event tab. It was working fine before it stopped working. I wonder if you mind having a look at it

File: models/Profile/Event.js [hunk 11]
```diff
@@ -34,6 +36,14 @@ const EventSchema = new Schema({
   price: {
     startingFrom: Number,
   },
+  userStatus: {
+    type: String,
```
Comment (dan-mba) [inline models/Profile/Event.js:40]: Why is this a string and not a boolean?
There are only 2 states:
attending & not speaking
attending & speaking

This is the perfect use case for a boolean

**Review by dan-mba — COMMENTED** [review]
I believe switching to a boolean will make for cleaner code.
The UI can then be improved by using the Toggle component for speaking.
@eddiejaoude Your thoughts?

File: models/Profile/Event.js (hunk 11 above)
Comment (eddiejaoude) [reply models/Profile/Event.js:40]: Could there be more? What about making it an object with ...?

- `isSpeaking`
- `isAttending`

Comment (dan-mba) [reply models/Profile/Event.js:40]: I was under the assumption that people were only posting events they are attending. That would make isAttending always true. Is this not the case?

Comment (eddiejaoude) [reply models/Profile/Event.js:40]: Yeah I think people are posting events they are aware of or want to support - which I think is a good thing, I am keen to have a central place for events

File: pages/account/manage/event/[[...data]].js [hunk 12]
```diff
@@ -102,19 +111,21 @@ export default function ManageEvent({ BASE_URL, event }) {
   const handleSubmit = async (e) => {
     e.preventDefault();
 
-    let alert = "created";
+let alert = "created";
```
Comment (dan-mba) [inline pages/account/manage/event/[[...data]].js:114]: please fix the indenting change

File: pages/account/manage/event/[[...data]].js [hunk 13]
```diff
@@ -263,10 +275,34 @@ export default function ManageEvent({ BASE_URL, event }) {
...
+                      <p className="text-sm text-primary-low-medium">
+                        For example: <i>The future of AI</i>
+                      </p>
+                    </div>
+                  )}
                   <div className="mt-1 sm:col-span-2 sm:mt-0">
                     <Input
                       type="number"
-                      min="0"
+                      minLength="0"
```
Comment (dan-mba) [inline pages/account/manage/event/[[...data]].js:305]: This should be `min` to prevent negative prices

**Review by dan-mba — CHANGES_REQUESTED** [review]

File: components/event/EventCard.js [hunk 14]
```diff
@@ -66,11 +66,11 @@ export default function EventCard({ manage, event, usernames }) {
...
                 </div>
-                {event.userStatus && (
+                {event.isSpeaking && (
                   <div className="text-primary-medium-low dark:text-primary-low-medium italic hidden lg:block">
-                    {event.userStatus}
-                    {event.userStatus == "speaking" && " at "} this event
-                    {event.userStatus == "speaking" && event?.speakingTopic && (
+                    {event.isSpeaking}
+                    {"Speaking at "} this event
+                    {event.isSpeaking  && event?.speakingTopic && (
```
Comment (dan-mba) [inline components/event/EventCard.js:73]: The `event.isSpeaking` check is unnecessary as this section is only displayed when this is true.

File: pages/api/system/reload.js [hunk 15]
```diff
@@ -261,6 +261,8 @@ export default async function handler(req, res) {
                   end: event.date.end,
                 },
                 url: event.url,
+                isSpeaking: event.isSpeaking,
```
[the remark this answers is not in the document]
Comment (akash19coder) [reply pages/api/system/reload.js:264]: @dan-mba  we don't have userStatus field... I am a bit confused why we are adding that?

**Review by dan-mba — APPROVED** [review]
Looks good to me

**Review by eddiejaoude — APPROVED** [review]
Great collaboration, thank you all 👍

**Pull request merged**